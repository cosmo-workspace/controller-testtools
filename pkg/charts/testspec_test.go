package charts

import (
	"io"
	"os"
	"testing"

	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"

	"github.com/jlandowner/helm-chartsnap/pkg/snap"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"

	unstructuredutil "github.com/jlandowner/helm-chartsnap/pkg/unstructured"
)

func TestSnapshot(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Testspec Suite")
}

var _ = Describe("ApplyDynamicFields", func() {
	load := func(filePath string) []unstructured.Unstructured {
		f, err := os.Open(filePath)
		Expect(err).NotTo(HaveOccurred())
		defer f.Close()

		buf, err := io.ReadAll(f)
		Expect(err).NotTo(HaveOccurred())

		manifests, err := unstructuredutil.Decode(string(buf))
		Expect(err).NotTo(HaveOccurred())

		return manifests
	}

	It("should replace specified fields", func() {
		testSpec := &TestSpec{
			DynamicFields: []ManifestPath{
				{
					APIVersion: "v1",
					Kind:       "Service",
					Name:       "chartsnap-app2",
					JSONPath: []string{
						"/spec/ports/0/targetPort",
					},
				},
				{
					APIVersion: "v2",
					Kind:       "Service",
					Name:       "chartsnap-app1",
					JSONPath: []string{
						"/spec/ports/0/targetPort",
					},
				},
				{
					APIVersion: "v1",
					Kind:       "service",
					Name:       "chartsnap-app1",
					JSONPath: []string{
						"/spec/ports/0/targetPort",
					},
				},
				{
					APIVersion: "v1",
					Kind:       "Service",
					Name:       "chartsnap-app1",
					JSONPath: []string{
						"/spec/ports/1/targetPort",
					},
				},
				{
					APIVersion: "apps/v1",
					Kind:       "Deployment",
					Name:       "chartsnap-app1",
					JSONPath: []string{
						"/metadata/labels/app.kubernetes.io~1version",
					},
				},
				{
					APIVersion: "v1",
					Kind:       "Pod",
					Name:       "chartsnap-app1-test-connection",
					JSONPath: []string{
						"/metadata/name",
					},
				},
			},
		}
		manifests := load("testdata/testspec_test.yaml")
		err := testSpec.ApplyFixedValue(manifests)
		Expect(err).NotTo(HaveOccurred())
		Expect(manifests).To(snap.MatchSnapShot())
	})
})
