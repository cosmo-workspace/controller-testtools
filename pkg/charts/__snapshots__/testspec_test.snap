['ApplyDynamicFields should replace specified fields 1']
SnapShot = """
[
  {
    \"apiVersion\": \"apps/v1\",
    \"kind\": \"Deployment\",
    \"metadata\": {
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"###DYNAMIC_FIELD###\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"chartsnap-app1\"
    },
    \"spec\": {
      \"replicas\": 1,
      \"selector\": {
        \"matchLabels\": {
          \"app.kubernetes.io/instance\": \"chartsnap\",
          \"app.kubernetes.io/name\": \"app1\"
        }
      },
      \"template\": {
        \"metadata\": {
          \"labels\": {
            \"app.kubernetes.io/instance\": \"chartsnap\",
            \"app.kubernetes.io/managed-by\": \"Helm\",
            \"app.kubernetes.io/name\": \"app1\",
            \"app.kubernetes.io/version\": \"1.16.0\",
            \"helm.sh/chart\": \"app1-0.1.0\"
          }
        },
        \"spec\": {
          \"containers\": [
            {
              \"image\": \"nginx:1.16.0\",
              \"imagePullPolicy\": \"IfNotPresent\",
              \"livenessProbe\": {
                \"httpGet\": {
                  \"path\": \"/\",
                  \"port\": \"http\"
                }
              },
              \"name\": \"app1\",
              \"ports\": [
                {
                  \"containerPort\": 80,
                  \"name\": \"http\",
                  \"protocol\": \"TCP\"
                }
              ],
              \"readinessProbe\": {
                \"httpGet\": {
                  \"path\": \"/\",
                  \"port\": \"http\"
                }
              },
              \"resources\": {},
              \"securityContext\": {}
            }
          ],
          \"securityContext\": {},
          \"serviceAccountName\": \"IyMjRFlOQU1JQ19GSUVMRCMjIw==\"
        }
      }
    }
  },
  {
    \"apiVersion\": \"v1\",
    \"kind\": \"Pod\",
    \"metadata\": {
      \"annotations\": {
        \"helm.sh/hook\": \"test\"
      },
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"1.16.0\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"###DYNAMIC_FIELD###\"
    },
    \"spec\": {
      \"containers\": [
        {
          \"args\": [
            \"chartsnap-app1:80\"
          ],
          \"command\": [
            \"wget\"
          ],
          \"image\": \"busybox\",
          \"name\": \"wget\"
        }
      ],
      \"restartPolicy\": \"Never\"
    }
  },
  {
    \"apiVersion\": \"v1\",
    \"kind\": \"Service\",
    \"metadata\": {
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"1.16.0\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"chartsnap-app1\"
    },
    \"spec\": {
      \"ports\": [
        {
          \"name\": \"http\",
          \"port\": 80,
          \"protocol\": \"TCP\",
          \"targetPort\": \"http\"
        },
        {
          \"name\": \"https\",
          \"port\": 443,
          \"protocol\": \"TCP\",
          \"targetPort\": \"###DYNAMIC_FIELD###\"
        }
      ],
      \"selector\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/name\": \"app1\"
      },
      \"type\": \"ClusterIP\"
    }
  },
  {
    \"apiVersion\": \"v1\",
    \"automountServiceAccountToken\": true,
    \"kind\": \"ServiceAccount\",
    \"metadata\": {
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"1.16.0\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"chartsnap-app1\"
    }
  }
]
"""

['LoadSnapshotConfig when loading .chartsnap.yaml should load config 1']
SnapShot = """
{
  \"DynamicFields\": [
    {
      \"Kind\": \"Secret\",
      \"APIVersion\": \"v1\",
      \"Name\": \"app1-cert\",
      \"JSONPath\": [
        \"/data/ca.crt\",
        \"/data/tls.crt\",
        \"/data/tls.key\"
      ],
      \"Base64\": true
    }
  ]
}
"""

['LoadSnapshotConfig when values.yaml has testSpec should load config 1']
SnapShot = """
{
  \"TestSpec\": {
    \"DynamicFields\": [
      {
        \"Kind\": \"Secret\",
        \"APIVersion\": \"v1\",
        \"Name\": \"app1-cert\",
        \"JSONPath\": [
          \"/data/ca.crt\",
          \"/data/tls.crt\",
          \"/data/tls.key\"
        ],
        \"Base64\": true
      }
    ]
  }
}
"""

['Merge should merge dynamic fields 1']
SnapShot = """
{
  \"DynamicFields\": [
    {
      \"Kind\": \"service\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1\",
      \"JSONPath\": [
        \"/spec/ports/0/targetPort\"
      ],
      \"Base64\": false
    },
    {
      \"Kind\": \"Pod\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1-test-connection\",
      \"JSONPath\": [
        \"/metadata/name\"
      ],
      \"Base64\": false
    },
    {
      \"Kind\": \"service\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1\",
      \"JSONPath\": [
        \"/spec/ports/0/targetPort\"
      ],
      \"Base64\": false
    },
    {
      \"Kind\": \"Service\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1\",
      \"JSONPath\": [
        \"/spec/ports/1/targetPort\"
      ],
      \"Base64\": false
    }
  ]
}
"""

['TestSpec ApplyDynamicFields should replace specified fields 1']
SnapShot = """
[
  {
    \"apiVersion\": \"apps/v1\",
    \"kind\": \"Deployment\",
    \"metadata\": {
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"###DYNAMIC_FIELD###\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"chartsnap-app1\"
    },
    \"spec\": {
      \"replicas\": 1,
      \"selector\": {
        \"matchLabels\": {
          \"app.kubernetes.io/instance\": \"chartsnap\",
          \"app.kubernetes.io/name\": \"app1\"
        }
      },
      \"template\": {
        \"metadata\": {
          \"labels\": {
            \"app.kubernetes.io/instance\": \"chartsnap\",
            \"app.kubernetes.io/managed-by\": \"Helm\",
            \"app.kubernetes.io/name\": \"app1\",
            \"app.kubernetes.io/version\": \"1.16.0\",
            \"helm.sh/chart\": \"app1-0.1.0\"
          }
        },
        \"spec\": {
          \"containers\": [
            {
              \"image\": \"nginx:1.16.0\",
              \"imagePullPolicy\": \"IfNotPresent\",
              \"livenessProbe\": {
                \"httpGet\": {
                  \"path\": \"/\",
                  \"port\": \"http\"
                }
              },
              \"name\": \"app1\",
              \"ports\": [
                {
                  \"containerPort\": 80,
                  \"name\": \"http\",
                  \"protocol\": \"TCP\"
                }
              ],
              \"readinessProbe\": {
                \"httpGet\": {
                  \"path\": \"/\",
                  \"port\": \"http\"
                }
              },
              \"resources\": {},
              \"securityContext\": {}
            }
          ],
          \"securityContext\": {},
          \"serviceAccountName\": \"IyMjRFlOQU1JQ19GSUVMRCMjIw==\"
        }
      }
    }
  },
  {
    \"apiVersion\": \"v1\",
    \"kind\": \"Pod\",
    \"metadata\": {
      \"annotations\": {
        \"helm.sh/hook\": \"test\"
      },
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"1.16.0\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"###DYNAMIC_FIELD###\"
    },
    \"spec\": {
      \"containers\": [
        {
          \"args\": [
            \"chartsnap-app1:80\"
          ],
          \"command\": [
            \"wget\"
          ],
          \"image\": \"busybox\",
          \"name\": \"wget\"
        }
      ],
      \"restartPolicy\": \"Never\"
    }
  },
  {
    \"apiVersion\": \"v1\",
    \"kind\": \"Service\",
    \"metadata\": {
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"1.16.0\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"chartsnap-app1\"
    },
    \"spec\": {
      \"ports\": [
        {
          \"name\": \"http\",
          \"port\": 80,
          \"protocol\": \"TCP\",
          \"targetPort\": \"http\"
        },
        {
          \"name\": \"https\",
          \"port\": 443,
          \"protocol\": \"TCP\",
          \"targetPort\": \"###DYNAMIC_FIELD###\"
        }
      ],
      \"selector\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/name\": \"app1\"
      },
      \"type\": \"ClusterIP\"
    }
  },
  {
    \"apiVersion\": \"v1\",
    \"automountServiceAccountToken\": true,
    \"kind\": \"ServiceAccount\",
    \"metadata\": {
      \"labels\": {
        \"app.kubernetes.io/instance\": \"chartsnap\",
        \"app.kubernetes.io/managed-by\": \"Helm\",
        \"app.kubernetes.io/name\": \"app1\",
        \"app.kubernetes.io/version\": \"1.16.0\",
        \"helm.sh/chart\": \"app1-0.1.0\"
      },
      \"name\": \"chartsnap-app1\"
    }
  }
]
"""

['TestSpec LoadSnapshotConfig when loading .chartsnap.yaml should load config 1']
SnapShot = """
{
  \"DynamicFields\": [
    {
      \"Kind\": \"Secret\",
      \"APIVersion\": \"v1\",
      \"Name\": \"app1-cert\",
      \"JSONPath\": [
        \"/data/ca.crt\",
        \"/data/tls.crt\",
        \"/data/tls.key\"
      ],
      \"Base64\": true
    }
  ]
}
"""

['TestSpec LoadSnapshotConfig when values.yaml has testSpec should load config 1']
SnapShot = """
{
  \"TestSpec\": {
    \"DynamicFields\": [
      {
        \"Kind\": \"Secret\",
        \"APIVersion\": \"v1\",
        \"Name\": \"app1-cert\",
        \"JSONPath\": [
          \"/data/ca.crt\",
          \"/data/tls.crt\",
          \"/data/tls.key\"
        ],
        \"Base64\": true
      }
    ]
  }
}
"""

['TestSpec Merge should merge dynamic fields 1']
SnapShot = """
{
  \"DynamicFields\": [
    {
      \"Kind\": \"service\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1\",
      \"JSONPath\": [
        \"/spec/ports/0/targetPort\"
      ],
      \"Base64\": false
    },
    {
      \"Kind\": \"Pod\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1-test-connection\",
      \"JSONPath\": [
        \"/metadata/name\"
      ],
      \"Base64\": false
    },
    {
      \"Kind\": \"service\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1\",
      \"JSONPath\": [
        \"/spec/ports/0/targetPort\"
      ],
      \"Base64\": false
    },
    {
      \"Kind\": \"Service\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1\",
      \"JSONPath\": [
        \"/spec/ports/1/targetPort\"
      ],
      \"Base64\": false
    }
  ]
}
"""
